src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -1277,67 +1277,83 @@ private static unsafe int ApplyMask(Span<byte> toMask, int mask, int maskIndex)
         {
             Debug.Assert(maskIndex < sizeof(int));
 
-            int maskShift = maskIndex * 8;
-            int shiftedMask = (int)(((uint)mask >> maskShift) | ((uint)mask << (32 - maskShift)));
-
-            // Try to use SIMD.  We can if the number of bytes we're trying to mask is at least as much
-            // as the width of a vector and if the width is an even multiple of the mask.
-            if (Vector.IsHardwareAccelerated &&
-                Vector<byte>.Count % sizeof(int) == 0 &&
-                toMask.Length >= Vector<byte>.Count)
+            fixed (byte* toMaskBeg = toMask)
             {
-                Vector<byte> maskVector = Vector.AsVectorByte(new Vector<int>(shiftedMask));
-                Span<Vector<byte>> toMaskVector = MemoryMarshal.Cast<byte, Vector<byte>>(toMask);
-                for (int i = 0; i < toMaskVector.Length; i++)
-                {
-                    toMaskVector[i] ^= maskVector;
-                }
+                byte* toMaskPtr = toMaskBeg;
+                byte* toMaskEnd = toMaskBeg + toMask.Length;
+                byte* maskPtr = (byte*)&mask;
 
-                // Fall through to processing any remaining bytes that were less than a vector width.
-                toMask = toMask.Slice(Vector<byte>.Count * toMaskVector.Length);
-            }
+                // use SIMD if possible.
 
-            // If there are any bytes remaining (either we couldn't use vectors, or the count wasn't
-            // an even multiple of the vector width), process them without vectors.
-            int count = toMask.Length;
-            if (count > 0)
-            {
-                fixed (byte* toMaskPtr = &MemoryMarshal.GetReference(toMask))
+                if (Vector.IsHardwareAccelerated && Vector<byte>.Count % sizeof(int) == 0 && toMask.Length >= Vector<byte>.Count)
                 {
-                    byte* p = toMaskPtr;
+                    // align our pointer to Vector<byte>.Count
+
+                    while ((long)toMaskPtr % Vector<byte>.Count != 0)
```

Comment:
This is potentially adding a bunch of mod operations... what does that do to the gains vectorization in this scenario provides?  Is it still worthwhile?

Reply:
Updated to operate unsigned -- optimizer translates this optimally to a `TEST toMaskPtr, 1Fh`.

Reply:
If we resolve the question about supporting unaligned loads from `Vector<T>`, you could do an unaligned load, masking out any bytes that occur after the first aligned address, which turns this into a single operation.